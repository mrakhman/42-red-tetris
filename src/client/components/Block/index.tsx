import { StyleSheet, View } from 'react-native';
import * as React from 'react';

import { cellState } from '/config/constants';

const styles = StyleSheet.create({
  outer: {
    alignItems: 'center',
    borderColor: '#879372',
    borderWidth: 1,
    height: 20,
    justifyContent: 'center',
    width: 20,
  },
  outerSpectrum: {
    alignItems: 'center',
    borderColor: '#879372',
    borderWidth: 1,
    height: 6,
    justifyContent: 'center',
    width: 6,
  },
  inner: {
    height: 12,
    width: 12,
    backgroundColor: '#879372',
  },
  innerSpectrum: {
    height: 2,
    width: 2,
    backgroundColor: '#879372',
  },
});

const Block = ({ value, isSpectrum }: { value: CellStateType | number, isSpectrum?: boolean }): JSX.Element => {
  return (
    isSpectrum
      ?
      <View style={[styles.outerSpectrum, value === cellState.OCCUPIED ? { borderColor: 'black' } : (value === cellState.BLOCKED ? { borderColor: 'black'} : undefined)]}>
        <View style={[styles.innerSpectrum, value === cellState.OCCUPIED ? { backgroundColor: 'black' } : (value === cellState.BLOCKED ? { backgroundColor: 'transparent'} : undefined)]} />
      </View>
      :
      <View style={[styles.outer, value === cellState.OCCUPIED ? { borderColor: 'black' } : (value === cellState.BLOCKED ? { borderColor: 'black'} : undefined)]}>
        <View style={[styles.inner, value === cellState.OCCUPIED ? { backgroundColor: 'black' } : (value === cellState.BLOCKED ? { backgroundColor: 'transparent'} : undefined)]} />
      </View>
  );
};

export default Block;
