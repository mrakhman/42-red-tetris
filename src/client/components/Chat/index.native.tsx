import { View } from 'react-native';

export const ChatWidget = View;
export const addResponseMessage = (text: string): string => text;
