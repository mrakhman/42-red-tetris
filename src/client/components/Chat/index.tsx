import 'react-chat-widget/lib/styles.css';
import { Widget as ChatWidget, addResponseMessage, isWidgetOpened, dropMessages } from 'react-chat-widget';

export {
  ChatWidget,
  addResponseMessage,
  isWidgetOpened,
  dropMessages,
};
