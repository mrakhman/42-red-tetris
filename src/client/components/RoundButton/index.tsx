import { TouchableOpacity, View, Text, ViewStyle } from 'react-native';
import React from 'react';

const RoundButton = ({
  size, color, label, style, onPress, disabled, text,
}: {
  size: number, color: string, label?: string,
  style?: ViewStyle, onPress?: () => void, disabled?: boolean, text?: string,
}): JSX.Element => {
  const button = React.useRef<TouchableOpacity>(null);

  return (
    <View style={[{ alignItems: 'center' }, style]}>
      <TouchableOpacity
        ref={button}
        disabled={disabled}
        onPress={() => onPress?.()}
        style={{
          backgroundColor: color,
          borderRadius: size / 2,
          height: size,
          width: size,
          shadowColor: 'rgba(0,0,0, .4)',
          shadowOffset: { height: 1, width: 1 },
          shadowOpacity: 20,
          shadowRadius: 5,
          justifyContent: 'center',
          alignItems: 'center',
        }} >
        {text && <Text style={{ fontSize: 25 }}>{text}</Text>}
      </TouchableOpacity>
      {label && <Text style={{ marginTop: 10 }}>{label}</Text>}
    </View>
  );
};

export default RoundButton;
