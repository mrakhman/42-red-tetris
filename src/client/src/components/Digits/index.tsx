import _ from 'lodash';
import { Asset } from 'expo-asset';
import { ImageBackground, View, ViewStyle } from 'react-native';
import * as React from 'react';

export default function Digits({ style }: { style?: ViewStyle }): JSX.Element {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const imageURI = Asset.fromModule(require('../../../assets/sprite.png')).uri;
  return (
    <View style={[{ flexDirection: 'row' }, style]} >
      {_.map(Array(6), (val, index: number) => <ImageBackground
        key={index}
        source={{ uri: imageURI }}
        style={{ width: 14, height: 24, overflow: 'hidden' }}
        imageStyle={{
          height: 186,
          left: -215,
          resizeMode: 'cover',
          top: -25,
          width: 400,
        }}
      />)}
    </View>
  );
}
