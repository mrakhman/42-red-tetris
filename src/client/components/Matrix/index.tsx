import _ from 'lodash';
import { View, ViewStyle, StyleSheet } from 'react-native';
import * as React from 'react';

import BlockComponent from '/client/components/Block';
import Block from '/client/models/block';
import { cellState } from '/client/constants/tetriminos.ts';

const blockWidth = 22;

const Matrix = ({ block, matrix, style }: { block: Block, matrix: Matrix, style?: ViewStyle }): JSX.Element => {
  return (
    <View style={[styles.container, { height: _.size(matrix) * blockWidth }, style]}>
      {_.map(matrix, (row, rowIndex) =>
        <View key={rowIndex} style={{ flexDirection: 'row', width: _.size(row) * blockWidth, justifyContent: 'space-between' }}>
          {_.map(row, (value, colIndex) => <BlockComponent key={colIndex} value={block.isBlock([rowIndex, colIndex]) ? cellState.OCCUPIED : value} />)}
        </View>)}
    </View>
  );
};

export default Matrix;

const styles = StyleSheet.create({
  container: {
    padding: 2, justifyContent: 'space-between', borderColor: 'black', borderWidth: 1, backgroundColor: '#9ead86',
  },
});
