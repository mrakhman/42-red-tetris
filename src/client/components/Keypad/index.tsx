import { View } from 'react-native';
import React, { useContext, Dispatch, SetStateAction, useState } from 'react';
import { StackNavigationProp } from '@react-navigation/stack';
import { useNavigation } from '@react-navigation/native';
import Sound from 'react-sound';

import { keyboard } from '/client/constants/keyboard';
import { SOCKETS } from '/config/constants';
import SocketContext from '/client/context/SocketContext';
import UserContext from '/client/context/UserContext';
import RoundButton from '/client/components/RoundButton';

type Props = {
  isPause?: boolean,
  setIsPause?: Dispatch<SetStateAction<boolean>>,
  opponentsNumber: number,
  isLeader?: boolean,
  gameStarted?: boolean,
  gameover?: boolean,
  isSoloMode?: boolean,
  speedMode?: boolean,
}

const Keypad = (props: Props): JSX.Element => {
  const { isPause, setIsPause, opponentsNumber, isLeader, gameStarted, gameover, isSoloMode, speedMode } = props;
  const socket = useContext(SocketContext);
  const [musicPlaying, setMusicPlaying] = useState(false);
  const { userContext, setUserContext } = useContext(UserContext);
  const navigation = useNavigation<StackNavigationProp<RootStackParamList, 'Playground'>>();
  const showStartButton = isLeader && !gameStarted;
  const showPauseButton = (isLeader && gameStarted);
  const isButtonDisabled = !isSoloMode && (gameover || opponentsNumber < 1);
  const showSpeedModeButton = speedMode !== undefined ? true : false; // Condition is true on Playground screen
  const showSoundButton = showSpeedModeButton;

  const socketEmitStartGame = () => {
    if (!socket) throw Error('No socket');
    socket.emit(SOCKETS.START_GAME, { username: userContext.username, roomName: userContext.room });
  };

  const socketEmitPlayPause = () => {
    if (isSoloMode) {
      if (setIsPause) setIsPause(prevState => !prevState);
    }
    if (!socket) throw Error('No socket');
    socket.emit(SOCKETS.PAUSE_GAME, { username: userContext.username, roomName: userContext.room });
  };

  const socketExitPage = () => {
    if (!socket) throw Error('No socket');
    setUserContext({username: undefined, room: undefined});
    if (setIsPause) setIsPause(true);
    setMusicPlaying(false);
    socket.emit(SOCKETS.PLAYER_LEFT, userContext.username);
    navigation.replace('Root');
  };

  const socketEmitSpeedMode = () => {
    if (!socket) throw Error('No socket');
    socket.emit(SOCKETS.SPEED_MODE, { username: userContext.username, roomName: userContext.room });
  };

  const keyDown = (key: number) => {
    // @ts-ignore https://github.com/microsoft/TSJS-lib-generator/pull/892
    document.dispatchEvent(new KeyboardEvent('keydown', { keyCode: key, which: key }));
  };
  return (
    <View style={{ width: 400 }}>
      <View style={{ width: '100%' }}>
        <View style={{ flexDirection: 'row', marginVertical: 20, justifyContent: 'space-between'}}>
          {showStartButton &&
          <RoundButton
            color={isButtonDisabled ? '#c0c0c0' : '#2dc421'} size={50} label='Start' text='â–¶'
            disabled={isButtonDisabled}
            onPress={() => socketEmitStartGame()}
          />}
          {showPauseButton &&
            <RoundButton
              disabled={isButtonDisabled}
              color={isButtonDisabled ? '#c0c0c0' : '#efcc19'} size={50} label={isPause ? 'Play' : 'Pause'} text={isPause ? 'â–¶' : '||' }
              onPress={() => socketEmitPlayPause()}
            />}
          {showSpeedModeButton &&
            <RoundButton
              disabled={isButtonDisabled}
              color={isButtonDisabled ? '#c0c0c0' : 'white'} size={50} label="Super speed" text={speedMode ? 'on' : 'off'}
              onPress={() => socketEmitSpeedMode()}
            />}
          {showSoundButton &&
            <RoundButton
              color={'white'} size={50} label="Sound" text={musicPlaying ? 'ðŸŽ¶' : 'ðŸ¤«' }
              onPress={() => setMusicPlaying(!musicPlaying)}
            />}
          <RoundButton
            color="white" size={50} style={{ margin: 0 }} label='Exit' text='â•³'
            onPress={() => socketExitPage()}
          />
        </View>
        <View style={{ flexDirection: 'row', alignItems: 'center', marginTop: 20 }}>
          <View style={{ marginRight: 20 }}>
            <RoundButton
              color={isButtonDisabled ? '#c0c0c0' : '#5a65f1'} size={160} label="Drop(Space)" style={{ marginTop: 20 }}
              disabled={isButtonDisabled || isPause}
              onPress={() => keyDown(keyboard.space)}
            />
          </View>
          <View style={{ marginLeft: 20 }}>
            <RoundButton
              color={isButtonDisabled ? '#c0c0c0' : '#5a65f1'} size={70} style={{ margin: 0 }}
              disabled={isButtonDisabled || isPause}
              onPress={() => keyDown(keyboard.rotate)}
            />
            <View style={{ flexDirection: 'row' }}>
              <RoundButton
                color={isButtonDisabled ? '#c0c0c0' : '#5a65f1'} size={70} style={{ margin: 0, marginRight: 60 }}
                disabled={isButtonDisabled || isPause}
                onPress={() => keyDown(keyboard.left)}
              />
              <RoundButton
                color={isButtonDisabled ? '#c0c0c0' : '#5a65f1'} size={70} style={{ margin: 0 }}
                disabled={isButtonDisabled || isPause}
                onPress={() => keyDown(keyboard.right)}
              />
            </View>
            <RoundButton
              color={isButtonDisabled ? '#c0c0c0' : '#5a65f1'} size={70} style={{ margin: 0 }}
              disabled={isButtonDisabled || isPause}
              onPress={() => keyDown(keyboard.down)}
            />
          </View>
        </View>
      </View>
      <Sound
        url={'https://www.soundhelix.com/examples/mp3/SoundHelix-Song-10.mp3'}
        playStatus={musicPlaying ? 'PLAYING' : 'STOPPED'}
        loop={true}
        autoLoad={true}
      />
    </View>
  );
};

export default Keypad;
