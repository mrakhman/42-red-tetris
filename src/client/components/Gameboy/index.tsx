import { StyleSheet, Text, View } from 'react-native';
import * as React from 'react';
import Keypad from '/client/components/Keypad';

import { useWindowDimensions } from '/client/hooks/useWindowDimensions';

export default function Gameboy({ children }: { children: JSX.Element[] | JSX.Element }): JSX.Element {
  const window = useWindowDimensions();
  const w = window.width;
  const h = window.height;
  const ratio = h / w;
  let scale;
  if (ratio < 1.5) {
    scale = h / 960;
  } else {
    scale = w / 560;
  }
  return (
    <View style={styles.container}>
      <View style={[styles.gameboy, { transform: [{ scale }] }]}>
        <Text style={styles.title}>R E D â–  T E T R I S</Text>
        <View style={styles.display}>
          {children}
        </View>
        <Keypad/>
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#009688',
    alignItems: 'center',
    justifyContent: 'center',
    padding: 20,
    height: '100%',
  },
  display: {
    alignItems: 'center',
    backgroundColor: '#9ead86',
    borderBottomColor: '#fa6b6b',
    borderLeftColor: '#980f0f',
    borderRightColor: '#fa6b6b',
    borderTopColor: '#980f0f',
    borderWidth: 5,
    height: 470,
    justifyContent: 'center',
    padding: 10,
    width: 400,
  },
  gameboy: {
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#af0000',
    borderRadius: 20,
    paddingBottom: 50,
    paddingHorizontal: 50,
  },
  title: {
    fontSize: 30,
    marginVertical: 20,
    textAlign: 'center',
  },
});
