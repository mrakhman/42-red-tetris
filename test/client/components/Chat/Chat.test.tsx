import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { ChatWidget } from '/client/components/Chat';

describe('Snapshots', () => {
  it('should render ChatWidget', () => {
    const props = {
      title: 'Chat title',
      subtitle: 'Chat subtitle',
      handleNewUserMessage: jest.fn(),
    };

    const wrapper = shallow(<ChatWidget {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
