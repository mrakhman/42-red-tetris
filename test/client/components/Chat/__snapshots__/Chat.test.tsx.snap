// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshots should render ChatWidget 1`] = `
<View
  handleNewUserMessage={[MockFunction]}
  subtitle="Chat subtitle"
  title="Chat title"
/>
`;
